Add Python Bindings for `SphericalJoints`

mstoelzle opened the pull request:
This pull request introduces support for spherical joints in the iDynTree Python bindings and improves the robustness and portability of Python tests. The changes include adding the `SphericalJoint` class to both SWIG and pybind11 bindings, updating test infrastructure to reliably locate built Python modules, and extending model parsing options to handle conversion between three revolute joints and a spherical joint. New tests are added to verify the correct behavior of spherical joints and model import/export options.

**Attention:** I needed the changes in commit `55594361fabf24c48314dd49f93fcab25b571742` to make the tests run on my installation, but I am also happy to remove these changes from the PR. In any case, I have confirmed that the bindings work.

**Spherical joint support:**

* Added `SphericalJoint` class to pybind11 bindings, exposing its constructor and methods (`set_joint_center`, `get_joint_center`) in `idyntree_model.cpp` and included the header in binding files. [[1]](diffhunk://#diff-bfaa82526d9a9bee9f0fb65f0ca33905678794468932350187c6cf64927b6fd3R13) [[2]](diffhunk://#diff-bfaa82526d9a9bee9f0fb65f0ca33905678794468932350187c6cf64927b6fd3R284-R289) [[3]](diffhunk://#diff-f251cb17f93b82eed5062a02aa7bfabfaa936275628fa45e83c457d4867e61f5R92) [[4]](diffhunk://#diff-f251cb17f93b82eed5062a02aa7bfabfaa936275628fa45e83c457d4867e61f5R271-R276)
* Added SWIG interface functions for `isSphericalJoint` and `asSphericalJoint` in `joints.i`. [[1]](diffhunk://#diff-08953427da3fd46453fbe7735e7e48ca5205f2dddae6c70ffe1ef33de984ba88R36-R42) [[2]](diffhunk://#diff-08953427da3fd46453fbe7735e7e48ca5205f2dddae6c70ffe1ef33de984ba88R63-R69)
* Added tests for `SphericalJoint` creation and behavior in both pybind11 and SWIG-based test suites. [[1]](diffhunk://#diff-f9ba380a9745c37715afd7f078af9cbae6a2d969875b9262b9a80d892d8cd783R201-R217) [[2]](diffhunk://#diff-01223d848f20d27e3cc63f06192cc53ffdaf97827a73043d892fbd45f4f899c5R29-R31)

**Model import/export enhancements:**

* Added options to export spherical joints as three revolute joints and to convert three revolute joints to a spherical joint in URDF model parsing/exporting, with associated pybind11 bindings for `ModelParserOptions` and `ModelExporterOptions`. [[1]](diffhunk://#diff-9610ee74986a792fd24f8f9c48515d1f7a2791b6765fc0bce2ca6397e68e3059L59-R75) [[2]](diffhunk://#diff-9610ee74986a792fd24f8f9c48515d1f7a2791b6765fc0bce2ca6397e68e3059L83-R102)
* Added tests to verify model loader behavior with and without conversion options using a URDF file compatible with spherical joint conversion. [[1]](diffhunk://#diff-f24b94b2a15742c2a90444033914c8b908569eca5fa7e66c12d047aea63271dbR3-R63) [[2]](diffhunk://#diff-f24b94b2a15742c2a90444033914c8b908569eca5fa7e66c12d047aea63271dbR114-R140)

**Python test infrastructure improvements:**

* Updated all Python tests to reliably locate and import freshly built modules by using `pathlib` and modifying `sys.path`, improving portability and robustness. [[1]](diffhunk://#diff-915a842f8648243cc1a8d66da5679b764840bfa5731a93b237bdded3b31ed1a5R3-R11) [[2]](diffhunk://#diff-d653d683eb783f14f4676f8216642a028205c5499e9d9415d65d6592c6c9a135R4-R12) [[3]](diffhunk://#diff-f9ba380a9745c37715afd7f078af9cbae6a2d969875b9262b9a80d892d8cd783R3-R11) [[4]](diffhunk://#diff-f24b94b2a15742c2a90444033914c8b908569eca5fa7e66c12d047aea63271dbR3-R63) [[5]](diffhunk://#diff-ae0e65291fd3f64638687a655093e5700359b6980456215c0214435ddf1aca6fR3-R11) [[6]](diffhunk://#diff-85fafc12978561a430f97ca8f2f54784c542236ab37f98fe7dfff2cedbefb18eR2-R11) [[7]](diffhunk://#diff-e6b6202991eda2759583c16ce9f4e80a6a674fe1ca68241d957ff3d6bff63beeL7-R15) [[8]](diffhunk://#diff-9687c1640b573646e548cdcc34252ee3bab476162f99ee4028065541fb4ffe51R6-R15) [[9]](diffhunk://#diff-926188a9da9c2584df45729fbc6bc1c68c581e7c8d6cfcc22d10b234a5609c31R3-R11) [[10]](diffhunk://#diff-01223d848f20d27e3cc63f06192cc53ffdaf97827a73043d892fbd45f4f899c5L7-R15) [[11]](diffhunk://#diff-5132152494856c3e57735276307270a8c1f9dfa57f20fc5058bad49c4a5cf667R2-R11)
* Set environment variables for test discovery to ensure correct module loading in CMake test configuration. [[1]](diffhunk://#diff-f0f7be184c517718801d49a84d3f0dcbc4972436ab027e92b01509f455bba396R5-R7) [[2]](diffhunk://#diff-6170c4ab38965ca5187a9d25501468ed1da7ef679da81e26c96f44a10866016dR7-R24)

**Test robustness:**

* Improved the inverse kinematics test to skip execution if required dependencies (IPOPT) are not available, preventing false negatives.

These changes collectively enhance the flexibility of the iDynTree Python bindings, especially for advanced joint modeling and URDF interoperability, while making the test suite more reliable and easier to run across different environments.

traversaro commented on bindings/python/tests/CMakeLists.txt:
```diff
@@ -4,9 +4,21 @@ add_custom_command(TARGET copy_python_tests_in_build PRE_BUILD
                    ${CMAKE_CURRENT_SOURCE_DIR} ${CMAKE_CURRENT_BINARY_DIR})
 
+set(IDYNTREE_PYTHON_TEST_PYTHONPATH
+    "PYTHONPATH=${CMAKE_BINARY_DIR}/bindings/python:${CMAKE_BINARY_DIR}/bindings/pybind11:${CMAKE_BINARY_DIR}/bindings:${CMAKE_BINARY_DIR}/lib/python:$ENV{PYTHONPATH}")
```
See https://github.com/robotology/idyntree/pull/1273/files#r2560193632 .

traversaro commented:
Thanks! Just to understand, did you passed the `iDynTree.i` to clang-format and that is where all the modifications come from? Probably we would need to add that to the check then (just a note to self, not something we need to do in this PR).

mstoelzle commented:
> Thanks! Just to understand, did you passed the `iDynTree.i` to clang-format and that is where all the modifications come from? Probably we would need to add that to the check then (just a note to self, not something we need to do in this PR).

I thought I would need to run the formatting as well on the `*.i` files that I had changed. However, when I did this, this introduced syntax errors. Therefore, I reverted the formatting of the `*.i`  and the formatting CI doesn't seem to care about it anyway.

traversaro commented:
@mstoelzle can I squash on merge?

flferretti commented:
@traversaro the pixi task only formats and checks `.cpp` and `.h` files (`git ls-files '*.cpp' '*.h' | parallel --will-cite -j 4 clang-format -i`). Maybe it should be extended to `.i` and similars
